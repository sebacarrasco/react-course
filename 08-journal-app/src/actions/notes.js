import { db } from "../firebase/firebase-config";
import { types } from "../types/types";
import { loadNotes } from '../helpers/loadNotes';
import Swal from 'sweetalert2';
import { fileUpload } from "../helpers/fileUpload";

// react-journal

export const startNewNote = () => {
    // Así es, aparte de dispatch se puede acceder a getState que
    // retorna un objeto con tooodos los estados en la store.
    // Bastante similar a useSelector (pero no recibe un callback)
    return async(dispatch, getState) => {

        const { uid } = getState().auth;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }
        try {
            // Dentro de collection va el path donde se quiere grabar
            const documentReference = await db.collection(`${ uid }/journal/notes`).add(newNote);
            dispatch(activeNote(documentReference.id, newNote));
            dispatch(addNewNote(documentReference.id, newNote));
        } catch (error) {
            console.log(error);
        }
    }
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = ( uid ) => {
    return async(dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: [...notes]
});


export const startSaveNote = (note) => {
    return async(dispatch, getState) => {

        const { uid } = getState().auth;

        // Si el url es undefined, entonces no lo mandaremos porque
        // firebase no deja guardar algo undefined (tira error)
        if (!note.url)
        {
            delete note.url;
        }

        // Eliminamos el id porque no lo guardamos dentro del objeto en firestore,
        // porque en firestore se llega al objeto a través del id, pero el objeto
        // en sí no contiene en sus propiedades el id (aunque podría tenerlo)
        const noteToFirestore = { ...note };
        delete noteToFirestore.id;

        await db.doc(`${ uid }/journal/notes/${ note.id }`).update(noteToFirestore);

        dispatch(refreshNote(note.id, note));
        Swal.fire('Saved', note.title, 'success');
    }
}

// Esta acción se encarga de actualizar en la store la note que se actualizó en firebase
export const refreshNote = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note
    }
})


export const startUploading = (file) => {
    return async (dispatch, getState) => {

        const { active:activeNote } = getState().notes;

        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            }
        });

        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;

        dispatch(startSaveNote(activeNote));

        Swal.close();
    }
}


export const startDeleting = (id) => {
    return async(disppatch, getState) => {

        const { uid } = getState().auth;
        await db.doc(`${ uid }/journal/notes/${ id }`).delete();

        disppatch(deleteNote(id));

    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
})

export const notesLogout = () => ({
    type: types.notesLogoutCleaning
})