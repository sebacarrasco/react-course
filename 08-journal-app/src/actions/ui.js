import { types } from "../types/types";


export const setRegisterError = (error) => ({
    type: types.uiSetRegisterError,
    payload: error
});

export const removeRegisterError = () => ({
    type: types.uiRemoveRegisterError
});


export const setLoginError = (error) => ({
    type: types.uiSetLoginError,
    payload: error
});

export const removeLoginError = () => ({
    type: types.uiRemoveLoginError
});


export const StartLoading = () =>({
    type: types.uiStartLoading
});

export const FinishLoading = () =>({
    type: types.uiFinishLoading
});