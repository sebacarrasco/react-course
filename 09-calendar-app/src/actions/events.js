import Swal from "sweetalert2";
import { fetchWithToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types";

export const eventStartAddNew = (event) => {
    return async(dispatch, getState) => {

        try {
            
            const response = await fetchWithToken('events', event, 'POST');
            const body = await response.json();

            const { uid, name } = getState().auth;

            if(body.ok)
            {
                event.id = body.event.id;
                event.user = {
                    _id: uid,
                    name
                }
                dispatch(eventAddNew(event));
            }

        } catch (error) {
            console.log(error);
        }

    }
}

export const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActiveEvent = () => ({
    type: types.eventClearActiveEvent
});

export const eventStartUpdate = (event) => {

    return async (dispatch) => {

        try {
            const response = await fetchWithToken(`events/${ event.id }`, event, 'PUT');
            const body = await response.json();
            if (body.ok)
            {
                dispatch(eventUpdated(event));
            }
            else
            {
                Swal.fire("Error", body.msg, "error");
            }
        } catch (error) {
            console.log(error);
        }
    }

}

export const eventUpdated = (event) => ({
    type: types.eventUpdated,
    payload: event
});

export const eventStartDelete = () => {

    return async (dispatch, getState) => {

        const { id } = getState().calendar.activeEvent;

        try {
            const response = await fetchWithToken(`events/${ id }`, {},  'DELETE');
            const body = await response.json();
            if (body.ok)
            {
                dispatch(eventDeleted());
            }
            else
            {
                Swal.fire("Error", body.msg, "error");
            }
        } catch (error) {
            console.log(error);
        }
    }

}

export const eventDeleted = () => ({
    type: types.eventDeleted
});

export const eventStartLoading = () => {
    return async(dispatch) => {

        try {
            const response = await fetchWithToken('events');
            const body = await response.json();
            // Pasamos las fechas de string a date
            const events = prepareEvents(body.events);
            dispatch(eventLoaded(events));
        } catch (error) {
            console.log(error);
        }

    }
}

export const eventLoaded = (events) => ({
    type: types.eventLoaded,
    payload: events
});

export const eventLogout = () => ({
    type: types.eventLogout
});