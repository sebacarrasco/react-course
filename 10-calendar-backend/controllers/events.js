const { response } = require('express');
const Event = require("../models/Event");

const getEvents = async(req, res = response) => {

    // El populate es para que no salga user: id_usuario,
    // sino que salga: user: { _id, name }
    // El id va siempre, pero así no mandamos la contraseña ni el mail.
    const events = await Event.find()
        .populate('user', 'name');

    res.json({
        ok: true,
        events
    });
}


const createEvent = async(req, res = response) => {

    const event = new Event(req.body);

    try {
        
        event.user = req.uid;

        const eventSaved = await event.save();
        res.status(201).json({
            ok: true,
            event: eventSaved
        });

    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: "Por favor hable con el administrador"
        });
    }
}


const updateEvent = async(req, res = response) => {

    const eventId = req.params.id;
    const { uid } = req;

    try {

        const event = await Event.findById(eventId);

        if (!event)
        {
            // 404 es de not found
            return res.status(404).json({
                ok: false,
                msg: "No existe un evento con ese id"
            });
        }

        if (event.user.toString() !== uid)
        {
            // 401 es cuando alguien trata de hacer algo que no está autorizado
            return res.status(401).json({
                ok: false,
                msg: "No tiene privilegio de editar este evento"
            });
        }

        const newEvent = {
            ...req.body,
            user: uid
        }
        
        // El new: true es para que retorne el evento actualizado
        const updatedEvent = await Event.findByIdAndUpdate(eventId, newEvent, { new: true });
        
        res.json({
            ok: true,
            event: updatedEvent
        });
        
    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: "Por favor hable con el administrador"
        });
    }

}


const deleteEvent = async(req, res = response) => {
    const eventId = req.params.id;
    const { uid } = req;

    try {

        const event = await Event.findById(eventId);

        if (!event)
        {
            // 404 es de not found
            return res.status(404).json({
                ok: false,
                msg: "No existe un evento con ese id"
            });
        }

        if (event.user.toString() !== uid)
        {
            // 401 es cuando alguien trata de hacer algo que no está autorizado
            return res.status(401).json({
                ok: false,
                msg: "No tiene privilegio de eliminar este evento"
            });
        }
        
        await Event.findByIdAndDelete(eventId);
        
        res.json({
            ok: true
        });
        
    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: "Por favor hable con el administrador"
        });
    }
}

module.exports = {
    getEvents,
    createEvent,
    updateEvent,
    deleteEvent
}