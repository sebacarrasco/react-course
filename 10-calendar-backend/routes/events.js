/**
Rutas para CRUD de Eventos
host + /api/events
*/

const express = require("express");
const { check } = require("express-validator");
const { getEvents, createEvent, updateEvent, deleteEvent } = require("../controllers/events");
const { isDate } = require("../helpers/isDate");
const { fieldValidator } = require("../middlewares/field-validator");
const { jwtValidator } = require("../middlewares/jwt-validator");


const router = express.Router();

// Como todas las rutas de este router usan el middleware jwtValidator,
// se puede "subir de nivel" para no ponerlo en cada una de las rutas por separado
router.use(jwtValidator);
// Todas las que están de acá para abajo usarán jwtValidator, si quiseramos que una
// ruta de este router no lo usara entonces basta con ponerla arriba de esto

router.get("/", getEvents);

router.post("/",
    [
        check("title", "title is required").not().isEmpty(),
        // Se pueden hacer validaciones personalizadas con express-validator
        check("start", "start date is required").custom(isDate),
        check("end", "end date is required").custom(isDate),
        fieldValidator
    ],
    createEvent
);

router.put("/:id",
    [
        check("title", "title is required").not().isEmpty(),
        // Se pueden hacer validaciones personalizadas con express-validator
        check("start", "start date is required").custom(isDate),
        check("end", "end date is required").custom(isDate),
        fieldValidator
    ],
    updateEvent
);

router.delete("/:id", deleteEvent);

module.exports = router;