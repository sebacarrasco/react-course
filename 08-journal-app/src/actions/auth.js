import Swal from 'sweetalert2';
import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { FinishLoading, StartLoading } from "./ui";
import { notesLogout } from './notes';

export const startLoginEmailPassword = (email, password) => {
    // Cuando una acción retorna un callback (una función)
    // el midleware la ejecutará. Además pasa como parámetro
    // el dispatch gracias a thunk
    return (dispatch) => {

        dispatch(StartLoading());

        // El return es solo para que en testing aparezcan las acciones login y finishLoading como
        // despachadas
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(FinishLoading());
            })
            .catch(error => {
                dispatch(FinishLoading());
                Swal.fire('Error', error.message, 'error');
            });
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        
        dispatch(StartLoading());

        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then( async({ user }) => {
                await user.updateProfile({displayName: name}); // Hay que establecerle el displayName a diferencia
                                                            // de cuando se hace la autenticación con google (se actualiza en db y el objeto)
                dispatch(login(user.uid, user.displayName));
                dispatch(FinishLoading());
            })
            .catch(error => {
                dispatch(FinishLoading());
                Swal.fire('Error', error.message, 'error');
            });
    }
}

export const startGoogleLogin = () => {
    // Como es una tarea asíncrona, se tiene que retornar una función.
    // Nótese que estas acciones son despachadas también.
    // Al ser despachadas y no retornar un objeto sino que una función,
    // entonces se ejecuta la función retornada y se pueden hacer dispatch dentro
    // de la misma función
    return (dispatch) => {
        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( userCred => {
                dispatch(login(userCred.user.uid, userCred.user.displayName))
            })
    }
    // El displayName lo tendremos siempre que se autentica con una red social, github o algo así
}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});


export const startlogout = () => {
    return async(dispatch) => {
        await firebase.auth().signOut(); //Solo puede fallar por conexión a internet
        dispatch(logout());
        dispatch(notesLogout());
    }
}

export const logout = () => ({
    type: types.logout
});