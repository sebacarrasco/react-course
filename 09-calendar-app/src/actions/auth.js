import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch"
import { types } from "../types/types";
import Swal from 'sweetalert2'
import { eventLogout } from "./events";


export const startLogin = (email, password) => {
    
    return async(dispatch) => {

        const response = await fetchWithoutToken('auth', { email, password }, 'POST');
        const body = await response.json();

        if (body.ok)
        {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        }
        else
        {
            Swal.fire('Error', body.msg, 'error');
        }
    }

}

export const startRegister = (email, password, name) => {

    return async(dispatch) => {

        const response = await fetchWithoutToken('auth/new', { email, password, name }, 'POST');
        const body = await response.json();

        if (body.ok)
        {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        }
        else
        {
            Swal.fire('Error', body.msg, 'error');
        }

    }

}

export const startChecking = () => {

    return async(dispatch) => {

        const response = await fetchWithToken('auth/renew');
        const body = await response.json();

        if (body.ok)
        {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        }
        else
        {
            dispatch(checkingFinish());
        }

    }

}

export const checkingFinish = () => ({
    type: types.authCheckingFinish
});

export const login = (user) => ({
    type: types.authLogin,
    payload: user
});

export const startLogout = () => {

    return (dispatch) => {
        localStorage.clear();
        dispatch(eventLogout());
        dispatch(logout());
    }

}

export const logout = () => ({
    type: types.authLogout
})