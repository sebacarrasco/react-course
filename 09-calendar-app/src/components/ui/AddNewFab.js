import React from 'react';
import { useDispatch } from 'react-redux';
import { uiOpenModal } from '../../actions/ui';

// Fab es de "Floating action button"

export const AddNewFab = () => {

    const dispatch = useDispatch();

    const handleClick = () => {
        dispatch(uiOpenModal());
    }

    return (
        <button
            className="btn btn-primary fab"
            onClick={ handleClick }
        >
                <i className="fas fa-plus"></i>

        </button>
    )
}
