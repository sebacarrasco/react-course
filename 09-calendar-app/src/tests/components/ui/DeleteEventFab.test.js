import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { DeleteEventFab } from '../../../components/ui/DeleteEventFab';
import { eventStartDelete } from '../../../actions/events';

jest.mock('../../../actions/events', () => ({
    eventStartDelete: jest.fn()
}))

// Config de la store
const middlewares = [ thunk ];
const mockStore = configureStore(middlewares);

const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store }>
        <DeleteEventFab />
    </Provider>
);

describe('Tests for <DeleteEventFab />', () => {
    
    test('should match the snapshot', () => {
        
        expect(wrapper).toMatchSnapshot();
        
    });

    test('should call eventStartDelete on click', () => {
       
        wrapper.find("button").simulate("click");
        expect(eventStartDelete).toHaveBeenCalled();

    });
    
    

});
