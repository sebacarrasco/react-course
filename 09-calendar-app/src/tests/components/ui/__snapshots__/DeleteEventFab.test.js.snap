// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <DeleteEventFab /> should match the snapshot 1`] = `
<button
  className="btn btn-danger fab-danger"
  onClick={[Function]}
>
  <i
    className="fas fa-trash"
  />
  <span>
     Borrar evento 
  </span>
</button>
`;
